### PR TITLE
Add raw sql support with execute

### PR DESCRIPTION
**Pull request by peterberkenbosch**
Add the support for executing raw sql from Repository classes. 

Some discusion is needed concerning the duplication, I would rather delegate the execute method from Repository to SqlAdapter.

### CODE REVIEW

**Comment by peterberkenbosch** [comment]
The failing spec is unrelated though, something in validations is breaking here
```
Lotus::Validations compatibility#test_0001_doesn't override already set accessor:
```

File: lib/lotus/repository.rb [hunk 1]
```diff
@@ -551,6 +551,42 @@ def transaction(options = {})
...
+      #
+      #   article = Article.new(title: 'Introducing transactions',
+      #     body: 'lorem ipsum')
+      #
+      #   resultset = ArticleRepository.execute("select * from articles")
+      #   resultset.each_hash do |result|
+      #     result # -> { id: 123, title: "Introducing transactions", body: "lorem ipsum"}
+      #   end
+      def execute(sql)
+        raise NotImplementedError unless @adapter.respond_to?(:execute)
```
Comment (jodosha) [inline lib/lotus/repository.rb]: @peterberkenbosch I'd get rid of this defensive line. Can you please add `#execute` to `Adapters::Abstract` and let it to raise this exception?

That class is a parent for the other two adapters that we support (memory and fs), they will inherit this behavior. Also, `Abstract` serves as a parent or official interface to support in order to be compliant with Lotus::Model.

File: lib/lotus/repository.rb [hunk 2]
```diff
@@ -551,6 +551,42 @@ def transaction(options = {})
...
+      #   end
+      #
+      #   article = Article.new(title: 'Introducing transactions',
+      #     body: 'lorem ipsum')
+      #
+      #   resultset = ArticleRepository.execute("select * from articles")
+      #   resultset.each_hash do |result|
+      #     result # -> { id: 123, title: "Introducing transactions", body: "lorem ipsum"}
+      #   end
+      def execute(sql)
```
Comment (jodosha) [inline lib/lotus/repository.rb]: @peterberkenbosch Probably `raw` is a better argument name. Please remember that at this level we are targeting any potential storage, so `sql` is a too specific term. :smile:

File: lotus-model.gemspec [hunk 3]
```diff
@@ -25,4 +25,5 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency 'bundler',  '~> 1.6'
   spec.add_development_dependency 'minitest', '~> 5'
   spec.add_development_dependency 'rake',     '~> 10'
+  spec.add_development_dependency 'pry-byebug', '3.1.0'
```
Comment (jodosha) [inline lotus-model.gemspec]: @peterberkenbosch We're trying to limit the amount of deps, we already have `byebug` in `Gemfile`. :smile_cat:

File: test/model/adapters/sql_adapter_test.rb [hunk 4]
```diff
@@ -1175,5 +1175,25 @@ class TestDeviceRepository
...
+      it 'returns the ResultSet from the executes sql' do
+        sql = "select * from users"
+        result = @adapter.execute(sql)
+        result.class.name.must_equal "SQLite3::ResultSet"
+        result.count.must_equal 3
+      end
+
+      it 'raises an exception when an invalid sql is provided' do
+        sql = "select foo from bar"
+        -> { @adapter.execute(sql) }.must_raise Sequel::DatabaseError
```
Comment (jodosha) [inline test/model/adapters/sql_adapter_test.rb]: @peterberkenbosch With #180 we have introduced `Lotus::Model::InvalidQueryError`. The long term goal is to let Lotus::Model to not raise adapters low level exceptions. Can you please catch `Sequel::DatabaseError` in the adapter and re-raise it? Thank you! :confetti_ball:

**Comment by jodosha** [comment]
@peterberkenbosch Thanks this PR. I'll take care now of that failing spec.

**Comment by jodosha** [comment]
@peterberkenbosch FYI - Build is green again :green_apple:

File: lib/lotus/model/adapters/abstract.rb [hunk 5]
```diff
@@ -201,6 +201,11 @@ def transaction(options = {})
         def connection_string
           raise NotSupportedError
         end
+
+        def execute(raw)
+          raise NotImplementedError
+        end
+
```
Comment (AlfonsoUceda) [inline lib/lotus/model/adapters/abstract.rb]: Can you remove this line?

File: lib/lotus/model/adapters/abstract.rb [hunk 6]
```diff
@@ -201,6 +201,11 @@ def transaction(options = {})
         def connection_string
           raise NotSupportedError
         end
+
+        def execute(raw)
```
Comment (AlfonsoUceda) [inline lib/lotus/model/adapters/abstract.rb]: Can you comment this method?

File: lib/lotus/model/adapters/sql_adapter.rb [hunk 7]
```diff
@@ -227,6 +227,19 @@ def connection_string
           Sql::Console.new(@uri).connection_string
         end
 
+        # Executes raw sql directly on the connection
+        #
+        # @return [Object]
```
Comment (AlfonsoUceda) [inline lib/lotus/model/adapters/sql_adapter.rb]: please add @param doc

File: lib/lotus/model/adapters/sql_adapter.rb [hunk 8]
```diff
@@ -227,6 +227,19 @@ def connection_string
...
+        #
+        # @since
```
Comment (AlfonsoUceda) [inline lib/lotus/model/adapters/sql_adapter.rb]: add @since x.x.x

File: lib/lotus/repository.rb [hunk 9]
```diff
@@ -551,6 +551,41 @@ def transaction(options = {})
         end
       end
 
+
```
Comment (AlfonsoUceda) [inline lib/lotus/repository.rb]: remove this line

File: lib/lotus/repository.rb [hunk 10]
```diff
@@ -551,6 +551,41 @@ def transaction(options = {})
...
+      # Executes the given raw sql on the adapter.
+      #
+      # Please note that it's only supported by some databases, a `NotImplementedError` will be
+      # raised when the adapter does not responds to the `execute` method.
+      #
+      # For advanced scenarios, please check the documentation of each adapter.
+      #
+      # @see Lotus::Model::Adapters::SqlAdapter#execute
+      #
+      # @since
```
Comment (AlfonsoUceda) [inline lib/lotus/repository.rb]: add @since x.x.x

File: lib/lotus/repository.rb [hunk 11]
```diff
@@ -551,6 +551,41 @@ def transaction(options = {})
...
+      #   end
+      #
+      #   article = Article.new(title: 'Introducing transactions',
+      #     body: 'lorem ipsum')
+      #
+      #   resultset = ArticleRepository.execute("select * from articles")
+      #   resultset.each_hash do |result|
+      #     result # -> { id: 123, title: "Introducing transactions", body: "lorem ipsum"}
+      #   end
+      def execute(raw)
```
Comment (AlfonsoUceda) [inline lib/lotus/repository.rb]: please add @param and @return doc

**Comment by peterberkenbosch** [comment]
Do you want these commits to be squashed @jodosha @joneslee85 @AlfonsoUceda ?

**Comment by AlfonsoUceda** [comment]
@peterberkenbosch IMO yes please ;) :+1: to PR

File: lib/lotus/repository.rb [hunk 12]
```diff
@@ -551,6 +551,43 @@ def transaction(options = {})
...
+      #   end
+      #
+      #   class ArticleRepository
+      #     include Lotus::Repository
+      #   end
+      #
+      #   article = Article.new(title: 'Introducing transactions',
+      #     body: 'lorem ipsum')
+      #
+      #   resultset = ArticleRepository.execute("select * from articles")
```
Comment (joneslee85) [inline lib/lotus/repository.rb]: what do you think about upcase SQL syntax?

File: test/model/adapters/sql_adapter_test.rb [hunk 13]
```diff
@@ -1175,5 +1175,25 @@ class TestDeviceRepository
...
+      before do
+        @adapter.create(collection, user1)
+        @adapter.create(collection, user2)
+        @adapter.create(collection, TestUser.new(name: 'S'))
+      end
+
+      it 'returns the ResultSet from the executes sql' do
+        raw = "select * from users"
+        result = @adapter.execute(raw)
+        result.class.name.must_equal "SQLite3::ResultSet"
```
Comment (joneslee85) [inline test/model/adapters/sql_adapter_test.rb]: what do you think about asserter `.must_be_kind_of`?

File: test/model/adapters/sql_adapter_test.rb [hunk 14]
```diff
@@ -1175,5 +1175,25 @@ class TestDeviceRepository
...
+        result.count.must_equal 3
+      end
+
+      it 'raises an exception when an invalid sql is provided' do
```
Comment (joneslee85) [inline test/model/adapters/sql_adapter_test.rb]: s/sql/SQL/g

**Comment by joneslee85** [comment]
Overall, :+1: from me

**Comment by jodosha** [comment]
@peterberkenbosch Can you please squash the commits, and I'll take care of the rest. Thank you :+1:

**Comment by peterberkenbosch** [comment]
@jodosha done :) And thanks.

**Pull request merged**